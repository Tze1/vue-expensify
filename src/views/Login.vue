<template>
  <div class="login component">
    <p class="login-byline">by <a href="http://tze1.com/" title="Tze-chiu Lei's portfolio site" target="_blank">Tze-chiu Lei</a></p>
    <p class="login-instructions">Please log in [no sign-up needed]:</p>
    <div class="login-btns">
      <button class="button login-btn third-party google" @click="startLoginGoogle" title="Log in with Google">Login with Google</button>
      <button class="button login-btn third-party facebook" @click="startLoginFacebook" title="Log in with Facebook">Continue with Facebook</button>
      <button class="button login-btn third-party twitter" @click="startLoginTwitter" title="Log in with Twitter">Login with Twitter</button>
      <button class="button login-btn third-party github" @click="startLoginGithub" title="Log in with GitHub">Login with GitHub</button>
    </div>
    <p class="login-credits">This DEMO app uses <a href="https://firebase.google.com/" target="_blank">Google Firebase</a> Authentication & Databases.<br />NOTE: If testing different sign-in providers, beware that each creates a separate account.  OAuth accounts-linking is beyond the scope of this demo.</p>
    <p class="login-appinfo">See source-code and unit-tests in the <a href="https://github.com/tze1/vue-expensify" target="_blank">GitHub project</a>.</p>
  </div>
</template>

<script>
  import { mapState } from 'vuex';
  import { firebase, googleAuthProvider, facebookAuthProvider, twitterAuthProvider, githubAuthProvider } from '../firebase/firebase';

	export default {
    name: 'Login',
    computed: mapState({
      user: state => state.auth.user,
    }),
    methods: {
      startLoginGoogle: function () {
        firebase.auth().signInWithPopup(googleAuthProvider).
          catch((err) => {
            // eslint-disable-next-line
            console.error('Firebase auth error: %s', err);
          });
      },
      startLoginFacebook: function () {
        firebase.auth().signInWithPopup(facebookAuthProvider).
          catch((err) => {
            // eslint-disable-next-line
            console.error('Firebase auth error: %s', err);
          });
      },
      startLoginTwitter: function () {
        firebase.auth().signInWithPopup(twitterAuthProvider).
          catch((err) => {
            // eslint-disable-next-line
            console.error('Firebase auth error: %s', err);
          });
      },
      startLoginGithub: function () {
        firebase.auth().signInWithPopup(githubAuthProvider).
          catch((err) => {
            // eslint-disable-next-line
            console.error('Firebase auth error: %s', err);
          });
      }
    }
  }
</script>

<style scoped lang="scss">
  @import '../styles/app.scss';

.login.component {
  .login-title {
    color: $red;
    margin-bottom: 0;
  }

  .login-byline {
    font-size: 1.2rem;
    margin-bottom: $space-sm;
    margin-top: -2.5rem;
  }

  .login-btns {
    display: flex;
    flex-flow: column nowrap;

    @include mq("tablet-wide") {
      flex-flow: row nowrap;
    }
  }

  .login-btn.third-party {
    background-color: transparent;
    background-position: center;
    background-repeat: no-repeat;
    border: none;
    color: transparent;
    height: 41px;
    margin: 0 auto $space-sm;
    overflow: visible;
    padding: 0;

    @include mq("tablet-wide") {
      margin: 0 2rem 0 0;
    }

    &:last-of-type {
      margin-bottom: 0;

      @include mq("tablet-wide") {
        margin-right: 0;
      }
    }

    &:hover {
      box-shadow: 0 0 1rem 0 #c66;
    }

    &.google {
      background-image: url('/img/btn-google-login.png');
      width: 186px;
    }

    &.facebook {
      background-image: url('/img/btn-facebook-login.png');
      width: 222px;
    }

    &.twitter {
      background-image: url('/img/btn-twitter-login.png');
      width: 176px;
    }

    &.github {
      background-image: url('/img/btn-github-login.png');
      width: 176px;
    }
  }

  .login-credits,
  .login-appinfo {
    font-size: 1.4rem;
  }
}
</style>
