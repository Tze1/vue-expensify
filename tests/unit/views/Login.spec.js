import { shallowMount } from '@vue/test-utils'
import sinon from 'sinon';
import Login from '@/views/Login'

describe('Login.vue', () => {
  const $store = {
    state: {
      auth: {
        user: undefined,
      },
    },
    getters: {
      userLoggedIn: state => {
        return (state.auth.user !== undefined);
      },
    },
  };
  const wrapper = shallowMount(Login, {
    mocks: {
      $store,
    },
  });
  const startLoginGoogleStub = jest.fn();
  const startLoginFacebookStub = jest.fn();
  const startLoginTwitterStub = jest.fn();
  const startLoginGithubStub = jest.fn();

  wrapper.setMethods({
    startLoginGoogle: startLoginGoogleStub,
    startLoginFacebook: startLoginFacebookStub,
    startLoginTwitter: startLoginTwitterStub,
    startLoginGithub: startLoginGithubStub,
  });

  it('renders Login view properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls startLoginGoogle properly when login-button is clicked', () => {
    wrapper.find('.login-btn.google').trigger('click');

    expect(startLoginGoogleStub).toHaveBeenCalled();
  });

  it('calls startLoginFacebook properly when login-button is clicked', () => {
    wrapper.find('.login-btn.facebook').trigger('click');

    expect(startLoginFacebookStub).toHaveBeenCalled();
  });

  it('calls startLoginTwitter properly when login-button is clicked', () => {
    wrapper.find('.login-btn.twitter').trigger('click');

    expect(startLoginTwitterStub).toHaveBeenCalled();
  });

  it('calls startLoginGithub properly when login-button is clicked', () => {
    wrapper.find('.login-btn.github').trigger('click');

    expect(startLoginGithubStub).toHaveBeenCalled();
  });
})

