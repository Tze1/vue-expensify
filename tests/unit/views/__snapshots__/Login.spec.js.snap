// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue renders Login view properly 1`] = `
<div class="login component">
  <p class="login-byline">by <a href="http://tze1.com/" title="Tze-chiu Lei's portfolio site" target="_blank">Tze-chiu Lei</a></p>
  <p class="login-instructions">Please log in [no sign-up needed]:</p>
  <div class="login-btns"><button title="Log in with Google" class="button login-btn third-party google">Login with Google</button> <button title="Log in with Facebook" class="button login-btn third-party facebook">Continue with Facebook</button> <button title="Log in with Twitter" class="button login-btn third-party twitter">Login with Twitter</button> <button title="Log in with GitHub" class="button login-btn third-party github">Login with GitHub</button></div>
  <p class="login-credits">This DEMO app uses <a href="https://firebase.google.com/" target="_blank">Google Firebase</a> Authentication &amp; Databases.<br>NOTE: If testing different sign-in providers, beware that each creates a separate account. OAuth accounts-linking is beyond the scope of this demo.</p>
  <p class="login-appinfo">See source-code and unit-tests in the <a href="https://github.com/tze1/vue-expensify" target="_blank">GitHub project</a>.</p>
</div>
`;
